src/flow/src/compute/render.rs

```diff
@@ -0,0 +1,467 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cell::RefCell;
+use std::collections::BTreeMap;
+use std::rc::Rc;
+
+use hydroflow::lattices::cc_traits::Get;
+use hydroflow::scheduled::graph::Hydroflow;
+use hydroflow::scheduled::graph_ext::GraphExt;
+use itertools::Itertools;
+use snafu::{OptionExt, ResultExt};
+
+use crate::adapter::error::{Error, EvalSnafu, InvalidQuerySnafu};
+use crate::compute::state::DataflowState;
+use crate::compute::types::{Arranged, Collection, CollectionBundle, ErrCollector, Toff};
+use crate::expr::{self, EvalError, GlobalId, LocalId, MapFilterProject, MfpPlan, ScalarExpr};
+use crate::plan::Plan;
+use crate::repr::{self, DiffRow, Row};
+use crate::utils::{ArrangeHandler, Arrangement};
+
+/// The Context for build a Operator with id of `GlobalId`
+pub struct Context<'referred, 'df> {
+    pub id: GlobalId,
+    pub df: &'referred mut Hydroflow<'df>,
+    pub compute_state: &'referred mut DataflowState,
+    /// a list of all collections being used in the operator
+    pub input_collection: BTreeMap<GlobalId, CollectionBundle>,
+    /// used by `Get`/`Let` Plan for getting/setting local variables
+    local_scope: Vec<BTreeMap<LocalId, CollectionBundle>>,
+    // Collect all errors in this operator's evaluation
+    err_collector: ErrCollector,
+}
+
+impl<'referred, 'df> Drop for Context<'referred, 'df> {
+    fn drop(&mut self) {
+        for bundle in std::mem::take(&mut self.input_collection)
+            .into_values()
+            .chain(
+                std::mem::take(&mut self.local_scope)
+                    .into_iter()
+                    .flat_map(|v| v.into_iter())
+                    .map(|(_k, v)| v),
+            )
+        {
+            bundle.collection.stream.drop(self.df);
+            drop(bundle.arranged);
+        }
+        // The automatically generated "drop glue" which recursively calls the destructors of all the fields (including the now empty `input_collection`)
+    }
+}
+
+// There is a false positive in using `Vec<ScalarExpr>` as key
+#[allow(clippy::mutable_key_type)]
+impl<'referred, 'df> Context<'referred, 'df> {
```

Comment:
how about adding methods to control collection bundles, e.g. `insert_global(global id, collection)`, `insert_local(scope, local id, collection)`